### DISCUSSION TITLE
Is it possible to communicate between agents using messages inside a host function?

### DISCUSSION THREAD

**Discussion by Himeshi** [FLAMEGPU/FLAMEGPU2#942] [Q&A] [answered]
Hi,

Is it possible to communicate between agents using messages inside a host function? This would be helpful to minimize the latency of data transfers between the host and device if, for example, there are several function calls that alternate between the host and the device, consecutively. 

Also, the documentation for FLAMEGPU2 states that "Simulation performance is significantly increased in comparison with CPU alternatives". Could you kindly point me to any experiments that verify this claim? Is there a study that compares the performance between the CPU and GPU for different number of agents?

Thanks!

**Comment by mondus** [comment, marked as answer]
Hi @Himeshi . I will answer these separately.

1) It is not possible for agents to read message on the host. Host functions are not designed for agent behaviour but rather for serial processes that may be required in between agent behaviour on the device. The host API does allow things like creation of agents and reading agent state but it is recommended to keep behaviour within device functions (as much as possible) to avoid the overhead of host to device communication. Another reason for this is that at some stage we plan on introducing a CPU back-end for FLAME as part of this the choice between device and host execution would be abstracted from users in line with our core design principle of "abstracting the implementation detail away form modellers".

2) The performance comparison against CPU agent based modelling libraries is anecdotal. You can view performance figures from other libraries and in fact run these yourself but a direct comparison would need a more careful benchmark exercise to ensure it was fair. We are currently conducting one and plan to release this as a blog post. An open access paper describing FLAME GPU 2 performance is also under review (since March).
